File: src/main/java/ski/mashiro/command/UserCommand.java
package ski.mashiro.command;

import net.mamoe.mirai.console.command.CommandSender;
import net.mamoe.mirai.console.command.FriendCommandSender;
import net.mamoe.mirai.console.command.SystemCommandSender;
import net.mamoe.mirai.console.command.java.JCompositeCommand;
import ski.mashiro.ScheduleQQAdvice;
import ski.mashiro.data.CourseData;
import ski.mashiro.data.UserData;
import ski.mashiro.file.UserFile;
import ski.mashiro.file.WhitelistFile;
import ski.mashiro.pojo.User;

import java.util.ArrayDeque;

/**
 * @author MashiroT
 */
public class UserCommand extends JCompositeCommand {
    public UserCommand() {
        super(ScheduleQQAdvice.INSTANCE, "user", "u");
        super.setPrefixOptional(true);
    }

    @SubCommand("bind")
    @Description("绑定用户")
    public void bind(CommandSender sender, String username, String apiToken) {
        if (hasNoPermission(sender)) {
            return;
        }
        if (UserFile.USERS_MAP.containsKey(sender.getUser().getId())) {
            sender.sendMessage("已经绑定了一个账户");
            return;
        }
        User rs = UserData.getUser(new User(sender.getUser().getId(), username, apiToken));
        if (rs == null) {
            sender.sendMessage("绑定失败");
            return;
        }
        rs.setApiToken(apiToken);
        UserFile.USERS.add(rs);
        UserFile.USERS_MAP.put(rs.getQq(), rs);
        CourseData.USER_TODAY_COURSE.putIfAbsent(rs.getQq(), new ArrayDeque<>());
        sender.sendMessage("绑定成功");
        if (CourseData.refreshQueue(sender.getUser().getId())) {
            sender.sendMessage("课程加载成功");
            return;
        }
        sender.sendMessage("Server Err");
    }

    @SubCommand("unbind")
    @Description("解除绑定")
    public void unbind(CommandSender sender) {
        if (hasNoPermission(sender) || hasNotBind(sender)) {
            return;
        }
        if (!UserFile.USERS_MAP.containsKey(sender.getUser().getId())) {
            sender.sendMessage("未绑定");
            return;
        }
        UserFile.USERS.removeIf(o -> o.getQq().equals(sender.getUser().getId()));
        UserFile.USERS_MAP.remove(sender.getUser().getId());
        sender.sendMessage("解绑成功");
        CourseData.USER_TODAY_COURSE.remove(sender.getUser().getId());
    }

    private boolean hasNoPermission(CommandSender sender) {
        if (!(sender instanceof SystemCommandSender)) {
            if (!(sender instanceof FriendCommandSender)) {
                return true;
            }
        }
        return !WhitelistFile.WHITELIST.contains(sender.getUser().getId());
    }

    private boolean hasNotBind(CommandSender sender) {
        return !UserFile.USERS_MAP.containsKey(sender.getUser().getId());
    }
}

File: src/main/kotlin/ski/mashiro/command/UserCommand.kt
package ski.mashiro.command

import net.mamoe.mirai.console.command.CommandSender
import net.mamoe.mirai.console.command.CompositeCommand
import net.mamoe.mirai.console.command.isConsole
import net.mamoe.mirai.console.command.isUser
import ski.mashiro.CourseScheduleQQAdvice
import ski.mashiro.data.UserData
import ski.mashiro.entity.User
import ski.mashiro.file.ConfigOp
import ski.mashiro.service.impl.UserServiceImpl

object UserCommand : CompositeCommand(
    CourseScheduleQQAdvice, "user", "u",
    description = "这是一个测试指令", // 设置描述，将会在 /help 展示
) {
    @net.mamoe.mirai.console.command.descriptor.ExperimentalCommandDescriptors
    override val prefixOptional: Boolean
        get() = true

    @SubCommand("bind", "b")
    suspend fun bindUser(sender: CommandSender, username: String, apiToken: String) {
        if (hasNoPerm(sender)) {
            return
        }
        if (username.length < 6 || apiToken.length < 32) {
            sender.sendMessage("非法参数")
            return
        }
        val senderQQ = sender.user!!.id
        if (UserData.userMap.contains(senderQQ)) {
            sender.sendMessage("不得重复绑定")
            return
        }
        if (!UserServiceImpl.verifyUser(username, apiToken)) {
            sender.sendMessage("用户名或token有误")
            return
        }
        UserData.userMap[senderQQ] = User(senderQQ, username, apiToken)
        sender.sendMessage("绑定用户: $username 成功")
    }

    @SubCommand("unbind", "ub")
    suspend fun unbindUser(sender: CommandSender) {
        if (hasNoPerm(sender)) {
            return
        }
        val senderQQ = sender.user!!.id
        if (!UserData.userMap.contains(senderQQ)) {
            sender.sendMessage("未绑定用户")
            return
        }
        val user = UserData.userMap.remove(senderQQ)
        sender.sendMessage("解绑用户: ${user!!.username} 成功")
    }

    private fun hasNoPerm(sender: CommandSender): Boolean {
        return sender.isConsole() || sender.isUser() &&
                ConfigOp.config.ownerQQ != sender.user.id &&
                !ConfigOp.config.whitelist.contains((sender.user.id))
    }
}